sycl/plugins/level_zero/pi_level_zero.hpp

```diff
@@ -833,13 +861,42 @@ struct _pi_program : _pi_object {
 
 struct _pi_kernel : _pi_object {
   _pi_kernel(ze_kernel_handle_t Kernel, pi_program Program)
-      : ZeKernel{Kernel}, Program{Program} {}
+      : ZeKernel{Kernel}, Program{Program}, MemAllocs{}, SubmissionsCount{0} {}
+
+  bool hasIndirectAccess() {
+    // Currently indirect access flag is set for all kernels and there is no API
+    // to check if kernel actually indirectly access smth.
+    return true;
+  }
 
   // Level Zero function handle.
   ze_kernel_handle_t ZeKernel;
 
   // Keep the program of the kernel.
   pi_program Program;
+
+  // If kernel has indirect access we need to make a snapshot of all existing
+  // memory allocations to defer deletion of these memory allocations to the
+  // moment when kernel execution has finished.
+  // We store iterators because iterator is not invalidated by insert/delete for
+  // std::map.
+  // Why need to take a snapshot instead of just reference-counting the
+  // allocations, because picture of active allocations can change during kernel
+  // execution (new allocations can be added) and we need to know which memory
+  // allocations were retained by this kernel to release them (and don't touch
+  // new allocations) at kernel completion.
+  std::list<std::map<void *, MemAllocRecord>::iterator> MemAllocs;
```

Comment:
What happens if the kernel is submitted multiple times (SubmissionsCount > 1)? The "retained" allocations at each submissions are different, right? How is it taken care of that some allocations need to be "released" after one kernel instance completion, and others after the others?

Reply:
I think I understand (after reviewing more code) that you add allocation at each submission, and release all of then when the submissions == 0, so it doesn't matter what individual submissions were retaining. Is that right? (maybe add a comment about that)

Reply:
You are right. It works as is, but I improved this by using unordered_set to not duplicate allocations for each submission.